lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java

```diff
@@ -21,16 +21,35 @@
 import java.util.Collection;
 import java.util.List;
 
+/**
+ * Abstracts a transactional backing store and staging area for {@link Notification notifications}.
+ *
+ * <p>A repository is responsible for handling CRUD and data parsing operations around these
+ * objects, in whatever scheme necessary.
+ *
+ * <p>Production implementations are expected to be thread safe, even across a network.
+ */
 public interface NotificationRepository {
 
     /**
-     * Retrieves the oldest {@code maxEvents} {@link Notification notifications}, oldest first.
+     * Retrieves the oldest {@code maxNotifications} {@link Notification notifications}, oldest
+     * first.
      *
      * <p>Once retrieved, a notification should not be retrieved again, atomically. That is, many
-     * threads looking up notifications at the same time should all et a unique non-overlapping sample
-     * of the oldest notifications. Subsequent calls should always return a unique set.
+     * threads looking up notifications at the same time should all et a unique non-overlapping
+     * sample of the oldest notifications. Subsequent calls should always return a unique set.
```

Comment:
There is no reason for batches to be large. Let this scale horizontally,
instead of trying to do a lot of things on a single node.

Lock calls are extremely lightweight *if* you acquire the lock the first
time you ask for it.

On Mon, Feb 1, 2016 at 1:26 PM, Alec Henninger <notifications@github.com>
wrote:

> In lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java
> <https://github.com/esbtools/event-handler/pull/8#discussion_r51475317>:
>
> >       *
> >       * <p>Once retrieved, a notification should not be retrieved again, atomically. That is, many
> > -     * threads looking up notifications at the same time should all et a unique non-overlapping sample
> > -     * of the oldest notifications. Subsequent calls should always return a unique set.
> > +     * threads looking up notifications at the same time should all et a unique non-overlapping
> > +     * sample of the oldest notifications. Subsequent calls should always return a unique set.
>
> Do you think it would be an excessive amount of lock calls to lock for
> each notification id? Notification batches will typically be large I think;
> we should process notifications as fast as lightblue will allow us.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/esbtools/event-handler/pull/8/files#r51475317>.
>